httpcore/_async/connection.py

```diff
@@ -99,7 +99,73 @@ def mark_as_ready(self) -> None:
             self.connection.mark_as_ready()
 
     async def start_tls(
-        self, hostname: bytes, timeout: Dict[str, Optional[float]] = None
+        self, hostname: bytes, timeout: Optional[TimeoutDict] = None
     ) -> None:
         if self.connection is not None:
             await self.connection.start_tls(hostname, timeout)
+
+
+class AsyncSOCKSConnection(AsyncHTTPConnection):
+    """An HTTP/1.1 connection with SOCKS proxy negotiation."""
+
+    def __init__(
+        self,
+        origin: Origin,
+        proxy_origin: Origin,
+        socks_version: str,
+        user_id: bytes = b"httpcore",
```

Comment:
This is required by SOCKS4.

SOCKS5 will require username/password and likely defining acceptable authentication methods. Which makes me think we might want separate `AsyncSOCKS4Connection` and `AsyncSOCKS5Connection`.